### DISCUSSION TITLE
Hot plug twice, than endpoint is always busy

### DISCUSSION THREAD

**Discussion by LYC584520** [hathach/tinyusb#1166] [Q&A]
MCU：STM32F405
TinyUSB：0.11.0
USB Device：CDC + HID
Log:
0x0803dff8 in usbd_edpt_xfer (rhport=1 '\001', ep_addr=0 '\000', buffer=0x0, total_bytes=0) at ../src/component/TinyUSB/src/device/usbd.c:1271
1271	  TU_ASSERT(_usbd_dev.ep_status[epnum][dir].busy == 0);

**Comment by ccrome** [comment]
Did you ever find the source of this problem?  I'm having the same assert and can't quite figure out what's going on.

I'm using CDC + UAC2.

I have found that commenting out a `tud_audio_buffer_and_schedule_control_xfer()` call from within a `tud_audio_get_req_entity_cb()` callback prevents this error.  But I'm not sure if that's a red herring, or if I'm onto something.

**Reply by ccrome** [reply]
:man_facepalming: I don't know how many different ways I can get burned by this same i.MX RT bug.  My problem is with the i.MX RT errata ERR050101 that reads:

> **ERR050101**
> USB: Endpoint conflict issue in device mode
> **Description**:
>An endpoint conflict occurs when the USB is working in device mode during an isochronous communication.
> When the endpointA IN direction is an isochronous IN endpoint, and the host sends an IN token to
> endpoint A on another device, then the OUT transaction may be missed regardless the OUT
> endpoint number. Generally, this occurs when the device is connected to the host through a hub and
> other devices are connected to the same hub.
> The affected OUT endpoint can be either control, bulk, isochronous, or an interrupt endpoint. After
> the OUT endpoint is primed, if an IN token to the same endpoint number on another device is
> received, then the OUT endpoint may be unprimed (cannot be detected by software), which causes
> this endpoint to no longer respond to the host OUT token, and thus, no corresponding interrupt
> occurs.
> **Workarounds**:
> Do not connect to a hub in case ISO IN endpoint(s) is used. When the hub(s) must be connected in
> this case, the endpoint number(s) of the ISO IN endpoint(s) should be different from the endpoint
> number(s) of any types of IN endpoint(s) used in any other device(s) connected to the same host.
> **Proposed Solution**:
> Fix in next revision

I was getting this endpoint busy error only on my 1060_evk board, and not on my Teensy 4.1.  That should have clued me in.  

Anyway, I'll leave this for anybody that searches for this in the future:
All i.MX RT chips that contain ERR050101 are irretrievably broken if you
   * use isochronous mode
   * and you have other USB devices on the same bus (*)

(*) They can be used in a closed system, where you have control over endpoint numbers of EVERY device on the bus, and manage to find a combination that doesn't conflict.

The list of devices that I know of that have the problem are
* [MIMXRT1015 REV A0](https://www.nxp.com/docs/en/errata/IMXRT1015CE.pdf)
* [MIMXRT1020 REV A0](https://www.nxp.com/docs/en/errata/IMXRT1020CE.pdf)
* [MIMXRT1024 REV A0](https://www.nxp.com/docs/en/errata/IMXRT1024CE.pdf)
* [MIMXRT1051 MIMXRT1052](https://www.nxp.com/docs/en/errata/IMXRT1050CE.pdf)  WILL NEVER BE FIXED
* [MIMXRT1061 MIMXRT1062 REV A](https://www.nxp.com/docs/en/errata/IMXRT1060CE_A.pdf)
* [MIMXRT1064 REV A](https://www.nxp.com/docs/en/errata/IMXRT1064CE_A.pdf)

The devices that DON'T have the problem
* i.MX RT1010
* [MIMXRT1015A1](https://www.nxp.com/docs/en/errata/IMXRT1015CE.pdf)
* [MIMXRT1020A1](https://www.nxp.com/docs/en/errata/IMXRT1020CE.pdf)
* [MIMXRT1024 REV A1](https://www.nxp.com/docs/en/errata/IMXRT1024CE.pdf)
* [MIMXRT1061 MIMXRT1062 REV B](https://www.nxp.com/docs/en/errata/IMXRT1060CE_B.pdf)
* [MIMXRT1064 REV B](https://www.nxp.com/docs/en/errata/IMXRT1064CE_B.pdf)
* [MIMXRT1070](https://www.nxp.com/docs/en/errata/IMXRT1170CE.pdf)
*

**Reply by hathach** [reply]
this is an interesting errata, I didn't know about this as well. Thanks for sharing

**Reply by ccrome** [reply]
I happened to have a i.MX RT 1062 rev B on hand, so I'm very proud of myself for swapping the rev A chip for the rev B, and it worked first try!  Watching enough repair videos on youtube gave me the confidence to try it.  Never tried swapping a 289 pin BGA before (or any BGA for that matter!)  Beginner's luck I guess.

**Reply by hathach** [reply]
That is a nightmare-level to rework an BGA. I have difficulty for just soldering the swd 0.5 pitch header

**Reply by ccrome** [reply]
**more flux** is usually the answer!
